Have PackageVariant set readiness gate on PackageRevisions

JamesMcDermott opened the pull request:
Implements [#615](https://github.com/nephio-project/porch/issues/610); documentation update in [docs PR nephio-project/nephio#204](https://github.com/nephio-project/docs/pull/204)

- incoming ReadinessGates and Conditions are now saved in PackageRevision creation flows
  - this includes init, clone, and edit operations
    - API-based init, clone, and edit operations (Create) will merge in any Conditions and ReadinessGates passed in in the request
      - any passed in in the API request are added to the defaults/existing ones
      - Conditions/ReadinessGates passed in in the API request will OVERRIDE default/existing ones
      - currently NO default Conditions/ReadinessGates are set - artifact of a reconsidered decision in response to a review comment
        - mechanism remains in place in case we later have need of setting some by default
- a successful pipeline run (defined as 1 or more runs in the same operation of the generictaskhandler.go::applyResourceMutations() function) sets the PackagePipelinePassed condition:
  - to False before starting
  - to True on successful completion
    - the patch operations for this appear in Git commit history, but NOT in the PackageRevision's task list to avoid polluting it
  - effectively preventing the PackageRevision from being proposed using `porchctl rpkg propose` while the pipeline is running
    - also, this validation is moved from porchctl into the API server to lock out API-based propose operations as well
      - also, this validation now checks which Conditions are unmet and provides a list to aid in resolving readiness situations
  - UNLESS operation is an update that does not change package resources
    - if this is the case, the pipeline is not run and, to save on expensive Git pushes, the PackagePipelinePassed condition is NOT set to False
- PackageVariant controller now sets its own PVOperationsComplete condition and readiness gate on PackageRevisions it manages
  - and manages this Condition independently to allow it to lock out propose operations while the controller is performing multiple operations
- PackageVariant controller can now handle optimistic concurrency conflicts when making its final (per-reconcile) status update to a PackageVariant


- refactors internal conversion of Kptfiles to YAML since readiness condition information is stored in the package Kptfile
  - unifies all cases to the same kyaml/yaml-based method (KptFile.ToYamlString() and ToYamlString(*fn.KubeObject))
  - this ensures consistency in the Kptfile YAML (indentation, field order etc.)
  - and reduces the chances of Git conflicts when setting and updating readiness conditions
- adds more info to error message in case of Git conflict when applying a patch
- renderPackageMutation mutation now has its own (arbitrary string) task type "render", to better distinguish it in task list and Git commit history
- CLI E2E tests now give more information on failure
  - pointing out config.yaml file containing command that produced unexpected result and 0-based index of the failed command in that file
- CLI E2E tests now fail on first test failure to allow easier identification of the failed test's output
- CLI E2E tests now respect a CLEANUP_ON_FAIL environment variable - if run with CLEANUP_ON_FAIL='false' and a CLI test fails, the suite will not clean that test's namespace off the Kind cluster, allowing for easier inspection of the state at time of failure
- some refactoring/renaming for readbility
- fixed typo in Makefile comment

JamesMcDermott committed:
Issue #615 - have PackageVariant set readiness gate on PackageRevisions

- PackageVariant controller now uses a readiness gate to allow
  a PackageRevision to complete its mutation pipeline before it
  is allowed to be proposed/approved
- refactored conversion of Kptfiles to YAML since readiness condition
  information is stored in the package Kptfile
  - unified all cases to the same kyaml/yaml-based method
    (KptFile.ToYamlString() and ToYamlString(*fn.KubeObject))
  - this ensures consistency in the YAML (indentation, field order etc.)
  - and reduces the chances of Git conflicts when setting and updating
    readiness conditions
- added more info to error message in case of Git conflict when applying
  a patch

https://github.com/nephio-project/nephio/issues/615

efiacor commented:
/test presubmit-nephio-go-test

efiacor commented on controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go:
```diff
@@ -334,6 +350,13 @@ func (r *PackageVariantReconciler) ensurePackageVariant(ctx context.Context,
 	if err = r.Client.Create(ctx, newPR); err != nil {
 		return nil, err
 	}
+
```
Looks good. I'd be interested to see how this works in a full test-infra e2e run.
Also, may be somewhat aligned with what @kispaljr was working on?

efiacor commented:
/test presubmit-nephio-go-test

JamesMcDermott commented:
/test presubmit-nephio-go-test

JamesMcDermott commented:
/test presubmit-nephio-go-test

JamesMcDermott commented:
/test presubmit-nephio-go-test

liamfallon commented:
/test presubmit-nephio-go-test

JamesMcDermott commented:
/test presubmit-nephio-go-test

JamesMcDermott committed:
Issue #615 - fixed readiness-gate Git conflicts

- previous solution broke existing functionality to copy comments between Kptfile objects
  - resulting in failing unit tests
    - fix for that brought back the Git conflicts caused by indentation, field order etc.
    - re-refactored conversion of Kptfiles to YAML
      - KubeObject-related conversion methods (KubeObject/YAML and YAML/KubeObject)
        now live out in the util package as they aren't Kptfile-specific any more

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
Issue #615 - adjusted test data missed in previous commit

https://github.com/nephio-project/nephio/issues/615

kispaljr commented on controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go:
```diff
@@ -56,8 +57,24 @@ type PackageVariantReconciler struct {
-	ConditionTypeReady   = "Ready"   // whether or not the reconciliation succeeded
+	ConditionTypeStalled        = "Stalled"                // whether or not the packagevariant object is making progress or not
+	ConditionTypeReady          = "Ready"                  // whether or not the reconciliation succeeded
+	ConditionTypePipelinePassed = "MutationPipelinePassed" // whether or not the mutation pipeline has completed successufully
```
I suggest to rename it to `PackagePipelinePassed` or similar, since the success of the validation pipeline should be checked, as well, not just the mutation pipeline.

kispaljr replied:
This of course also applies to all other places where it is referenced as a "mutation pipeline" (e.g. in conditions)

JamesMcDermott replied:
Done

kispaljr commented on pkg/kpt/api/kptfile/v1/types.go:
```diff
@@ -210,6 +233,8 @@ type PackageInfo struct {
 	// Relative slash-delimited path to the license file (e.g. LICENSE.txt)
 	LicenseFile string `yaml:"licenseFile,omitempty" json:"licenseFile,omitempty"`
 
+	ReadinessGates []ReadinessGate `yaml:"readinessGates,omitempty" json:"readinessGates,omitempty"`
```
This change make kpt packages managed by porch incompatible with kpt packages by kpt, whose Kptfile schema is defined here: https://github.com/kptdev/kpt/blob/main/pkg/api/kptfile/v1/types.go

kispaljr replied:
I wasn't able to attend to the latest meetings, do we have a community agreement/decision to break this compatibility?

efiacor replied:
On this, why are we using/defining our own ktp api?
There is similar kpt related stuff in the [nephio repo ](https://github.com/nephio-project/nephio/tree/main/krm-functions/lib/kptfile/v1).
Surely we should try to avoid customising an internal version and look to having a single shared api.

kispaljr reviewed (commented):
I welcome this PR, and wholeheartedly agree with the general direction. :)
These are my initial comments, but could you give me also a couple more days to understand the logic a bit better?

kispaljr commented on controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go:
```diff
@@ -347,6 +371,29 @@ func (r *PackageVariantReconciler) ensurePackageVariant(ctx context.Context,
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: newPR.GetName(), Namespace: newPR.GetNamespace()}, &refreshedPR); err != nil {
+		return nil, err
+	}
+	newPR.ResourceVersion = refreshedPR.ResourceVersion
```
Why are only ResourceVersion and Tasks are copied and every other potential changes are reverted back?
Why aren't we simply Get() the PR into &newPR, set the condition and update?

JamesMcDermott replied:
Done - the copy/revert thing with refreshedPR was left over from solving an earlier hurdle and it isn't needed now.

kispaljr commented on controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go:
```diff
@@ -347,6 +371,29 @@ func (r *PackageVariantReconciler) ensurePackageVariant(ctx context.Context,
 		}
 	}
 
+	// TODO: remove if OK to exclude PackageRevision from cache of r.Client
```
remove this comment if not needed

JamesMcDermott replied:
Done

kispaljr commented on controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go:
In the findAndUpdateExistingRevisions() method, at line 473:
```			// Save the updated PackageRevisionResources
			if err := r.updatePackageResources(ctx, prr, pv); err != nil {
				return nil, err
			}
```
The pipeline will be executed here, as well, so the Condition should be set to true or false here, as well, based on the results.

JamesMcDermott replied:
Done both:
- the `findAndUpdateExistingRevisions()` one could be a bit better by including the initial set of the Condition directly in `prr` instead of doing a separate `r.Client.Update()` - I want to look into automated (and a few more manual) tests first though
- the `updateDraft()` one took more time to work around to avoid ending up in an endless loop where doing the `Update()` to set the Condition would trigger the pipeline, resulting in a rendering result which we'd need to record in the Condition by doing an `Update()`, triggering the pipeline... you get the idea. :)
  - thanks to @liamfallon for suggesting we have the server detect and skip the pipeline when an update only sets the Condition/readiness info
  - works as desired when creating a PackageVariant - unable to test in this update flow due to immediate Git conflicts as seen in [#469](https://github.com/nephio-project/nephio/issues/469)

JamesMcDermott committed:
Issue #615 - skip pipeline on PackageRevision update if only updating pipeline-passed condition

- review comments to increase applicability of pipeline-passed readiness condition
  - to enable this: if an incoming PackageRevision update changes only that
    readiness condition, detect that, skip the pipeline/rendering/etc., and
    only update the Kptfile with the readiness condition update
- other tidy-up/wording comments

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
Issue #615 - added end-to-end test for PackageVariant readiness gate

 - also broke ground on testing PackageVariants in end-to-end tests at all

 https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
Issue #615 - added end-to-end test for PackageVariant readiness gate

 - also broke ground on testing PackageVariants in end-to-end tests at all

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
Issue #615 - removed missed line that locked tests' namespace to 'porch'

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
Issue #615 - E2E tests - additional repo cleanup for package variants

- tests now delete package variants whose upstream is a repo created by the test

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
Issue #615 - wired readiness info through into PackageRevision creation flows

- this includes both init and clone operations
  - a new PackageRevision now has the PackagePipelinePassed Condition and
    ReadinessGate by default
  - set to False on creation, but the initial (no-op) render run effectively
    sets it to True by default
  - API-based init operations (Create) will include any Conditions passed in
    in the request
    - and add ReadinessGates based on the Conditions
- a successful render run sets the PackagePipelinePassed condition:
  - to False before starting
  - to True on successful completion
  - effectively preventing the PackageRevision from being proposed using porchctl
    rpkg propose while the pipeline is running
    - a future commit will move this validation from porchctl into the API server
      to lock out API-based propose operations as well
- PackageVariant controller sets its own condition and readiness gate on
  PackageRevisions it manages
  - and manages this Condition independently to allow it to lock out propose
    operations while the controller is performing multiple operations
- PackageVariant controller can handle optimistic concurrency conflicts when
  making its final (per-reconcile) status update to a PackageVariant
- some refactoring/renaming for readbility
- fixed typo in Makefile comment

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
Issue #615 - unit test fix

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
Issue #615 - move readiness validation for propose and approve into server

- apply readiness-gate lockout to API operations as well as porchctl CLI

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
fix word in comment that got caught in variable name refactor

liamfallon commented on pkg/engine/engine.go:
```diff
@@ -516,6 +535,35 @@ func (cad *cadEngine) UpdatePackageResources(ctx context.Context, repositoryObj
 	return repoPkgRev, renderStatus, nil
 }
 
+func pushPipelineReadinessGate(ctx context.Context, repo repository.Repository, repoPr repository.PackageRevision) error {
```
This makes the engine handling of drafts even more complex but there is no oopter way to do this with the current code structure. Let's refactor engine.go in Nephio R5.

liamfallon reviewed (commented):
/approve

Very comprehensive work @JamesMcDermott

JamesMcDermott committed:
Issue #615 - update copyright comments to 2025

https://github.com/nephio-project/nephio/issues/615

lapentad commented on pkg/engine/engine.go:
```diff
@@ -191,12 +191,12 @@ func (cad *cadEngine) CreatePackageRevision(ctx context.Context, repositoryObj *
-	pkgRevMeta, err = cad.metadataStore.Create(ctx, pkgRevMeta, repositoryObj.Name, repoPkgRev.UID())
+	pkgRevMeta, err = cad.metadataStore.Create(ctx, pkgRevMeta, apiRepo.Name, repoPkgRev.UID())
 	if err != nil {
 		if (apierrors.IsUnauthorized(err) || apierrors.IsForbidden(err)) && repository.AnyBlockOwnerDeletionSet(obj) {
```
```suggestion
		if (apierrors.IsUnauthorized(err) || apierrors.IsForbidden(err)) && repository.AnyBlockOwnerDeletionSet(newPkgRev) {
```

lapentad commented on pkg/engine/engine.go:
```diff
@@ -191,12 +191,12 @@ func (cad *cadEngine) CreatePackageRevision(ctx context.Context, repositoryObj *
-	pkgRevMeta, err = cad.metadataStore.Create(ctx, pkgRevMeta, repositoryObj.Name, repoPkgRev.UID())
+	pkgRevMeta, err = cad.metadataStore.Create(ctx, pkgRevMeta, apiRepo.Name, repoPkgRev.UID())
 	if err != nil {
 		if (apierrors.IsUnauthorized(err) || apierrors.IsForbidden(err)) && repository.AnyBlockOwnerDeletionSet(obj) {
```
```suggestion
		if (apierrors.IsUnauthorized(err) || apierrors.IsForbidden(err)) && repository.AnyBlockOwnerDeletionSet(newPkgRev) {
```
I think this slipped during the rebase

JamesMcDermott replied:
Yep - thanks. Fixed there

kispaljr commented:
I read through the latest code changes. All my previous comments are now resolved.
I focused my review on the `pkg/task` package, and that seems excellent to me.

all-in-all: it looks good to me, it can be merged from my part

kispaljr commented on pkg/task/clone.go:
```diff
@@ -92,6 +94,14 @@ func (m *clonePackageMutation) apply(ctx context.Context, resources repository.P
+		if file.Status == nil {
+			file.Status = &kptfile.Status{}
+		}
+		file.Status.Conditions = slices.Concat(file.Status.Conditions, kptfile.ConvertApiConditions(m.pkgRev.Status.Conditions))
```
It's a minor thing, but I would expect the Conditions and readinessGates coming from the Kptfile and from the PackageRevision object to be merged together, instead of concatenated. Basically I would expect duplicate ConditionTypes to be removed from both merged lists.

kispaljr commented on pkg/task/init.go:
```diff
@@ -49,21 +54,27 @@ func (m *initPackageMutation) apply(ctx context.Context, resources repository.Pa
 		return repository.PackageResources{}, nil, err
 	}
+
+	readinessConditions := slices.Concat(defaultConditions, m.pkgRev.Status.Conditions)
```
Same as with "clone" I expect a merging of the two lists by ConditionType, as opposed to concatenations

kispaljr reviewed (commented):
Also, as we already discussed on Slack, it would make porch behavior  more consistent if the readinessGates + conditions would be also merged from the PR objects that are created with an edit task (similarly how you implemented it to clone and init), but that is not mandatory. This PR can be merged without that.

efiacor commented:
Need to run it through a few e2e runs. Hit an issue on the free5gc suite. Will run again and collect data if it's a recurring issue.

JamesMcDermott committed:
Issue #615 - review comment fixes

- Edit operation now includes ReadinessGates and Conditions passed
  in as part of the edit request
  - adding them to existing ones from the Kptfile in the
    PackageRevision being edited
- Init, Edit, and Clone operations now merge their lists of
  Conditions/ReadinessGates (any passed in in the API request
  are added to the defaults/existing ones)
  - instead of using slices.Concat which allowed duplicates
    (with same Condition.Type) to stay
  - Conditions/ReadinessGates passed in in the API request
    will OVERRIDE default/existing ones

https://github.com/nephio-project/nephio/issues/615

liamfallon commented:
Just waiting on @efiacor, are the e2e issues resolved?

JamesMcDermott committed:
Issue #615 - unify sorting on Condition and ReadinessGate lists

- alleviates issue where config injection from PackageVariant
  resulted in both PackageVariant controller detecting a Kptfile
  change (and updating the PackageRevision with a new patch task)
  and the API server NOT detecting any Kptfile change, resulting
  in the server attempting to apply a nil patch mutation and
  massive panics

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
Issue #156 - unify sorting on Condition and ReadinessGate lists

- alleviates issue where config injection from PackageVariant
  resulted in both PackageVariant controller detecting a Kptfile
  change (and updating the PackageRevision with a new patch task)
  and the API server NOT detecting any Kptfile change, resulting
  in the server attempting to apply a nil patch mutation and
  massive panics

https://github.com/nephio-project/porch/pull/156

JamesMcDermott committed:
Small Sonar fix in touched clone_test

JamesMcDermott committed:
merge conflict fix

JamesMcDermott committed:
Issue #615 - further fixes for issues in free5gc e2e suite

- adjustments to usage of Kubernetes IO annotation
  "internal.config.kubernetes.io/path" to keep it present
  consistently and avoid it showing up as a diff in update
  operations
- CLI E2E tests now respect a CLEANUP_ON_FAIL variable -
  if run with CLEANUP_ON_FAIL='false' and a CLI test fails,
  the suite will not clean that test's namespace off the Kind
  cluster, allowing for easier inspection of the state at
  at time of failure

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
conflict resolution compile fixes

JamesMcDermott committed:
Skip kioutil legacy annotation linting error

JamesMcDermott committed:
Skip kioutil legacy annotation linting error

JamesMcDermott committed:
Issue #615 - more fixes for issues in free5gc e2e suite

- in free5gc suite, one deployment package requires Kubernetes
  IO annotation "internal.config.kubernetes.io/index" before
  Nephio approval controller will take it in hand
  - adjusted usage to avoid removing that annotation as well
  - looks like we get away with clearing the legacy
    annotations, at least
- when updating package revision resources, skip closing
  pipeline readiness gate if no actual resource change
  (new resources are equal to old resources)
- add text to update where we commit/push to close pipeline
  readiness gate before render - better visibility in Git
  history

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
Issue #615 - more fixes for issues in free5gc e2e suite

- pushing code only
- tests likely to fail; adjustments pending confirmation that
  this still works with free5gc suite

JamesMcDermott committed:
Issue #615 - final fix for free5gc suite

- PackagePipelinePassed readiness condition is now set to
  "True" in genericTaskHandler, at the last possible moment
  before returning to cadEngine
  - unified approach for consistency
  - allows us to be sure it covers all mutations in any given
    create/update operation as some operations/packages may
    include (for example) multiple render mutations
  - also means lock/unlock commits show up as distinct commits
    in Git history
    - but specifically excluded them from task lists
- edit and clone operations now include default readiness
  gates and conditions
- gave render mutation its own (arbitrary string) task type
  "render", to better distinguish it in task list and Git
  commit history

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott commented:
/retest

JamesMcDermott committed:
Issue #615 - final fix to ensure final render task

- final render mutation should now run only on actual changes
  to resources
  - and not run on empty updates, preserving state of task list

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
Issue #615 - revert Kubernetes IO annotation handling changes

- further testing revealed annotations were a red herring
  - real issue lay in readiness-info patches not triggering
    Kpt package render operations

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
update some missed copyright comments

JamesMcDermott commented:
> Just waiting on @efiacor, are the e2e issues resolved?

Investigated and resolved the issues the e2e tests were running into - can confirm passing runs of both the free5gc and oai suites on latest commit to this PR.

JamesMcDermott committed:
Untangle import cycles caused by kpt-functions-sdk moving in

JamesMcDermott committed:
initial coverage increases

- refactored package revision variable names in edit mutation
  - had old and new the wrong way round

JamesMcDermott commented:
/retest

JamesMcDermott committed:
more coverage on edit.go

JamesMcDermott committed:
more coverage

efiacor commented on controllers/main.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 The kpt and Nephio Authors
+// Copyright 2022, 2025 The kpt and Nephio Authors
```
If this is just a new line we can omit the licence change

efiacor commented on controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go:
```diff
@@ -50,14 +51,36 @@ func (o *Options) BindFlags(_ string, _ *flag.FlagSet) {}
 type PackageVariantReconciler struct {
 	client.Client
+
+	// Reader goes directly to the Kubernetes API server and
```
Not sure this comment is needed

JamesMcDermott replied:
OK - removed. Might have been just my unfamiliarity with the package that made me think it was needed.

efiacor commented on controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go:
```diff
@@ -50,14 +51,36 @@ func (o *Options) BindFlags(_ string, _ *flag.FlagSet) {}
-	ConditionTypeReady   = "Ready"   // whether or not the reconciliation succeeded
+	ConditionTypeStalled            = "Stalled"              // whether or not the packagevariant object is making progress
+	ConditionTypeReady              = "Ready"                // whether or not the reconciliation succeeded
+	ConditionTypeAtomicPVOperations = "PVOperationsComplete" // whether or not the package's pipeline has completed successfully
```
PipelineExecutionsComplete?

JamesMcDermott replied:
Good catch, though it's actually the comment that needs updated here - this condition type is managed by the PackageVariant controller to enclose "any set of operations by the PackageVariant on a PackageRevision". I've updated the comment.

efiacor commented on controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go:
```diff
@@ -50,14 +51,36 @@ func (o *Options) BindFlags(_ string, _ *flag.FlagSet) {}
+	// Reader goes directly to the Kubernetes API server and
+	// may be used for Get and List operations if you need to
+	// bypass the cache in Client (which may be stale).
+	client.Reader
```
Can we name the Client and Reader and ref them using the name instead?

JamesMcDermott replied:
Names such as what, and what would that achieve? (honest question/suggestion-elicitation) (discuss?)
If readability, surely something like r.Client.Get or r.Reader.List is clearer than, say, r.c.Get or r.r.List...
Also, this is a fairly broad refactoring that will touch another 25+ references in packagevariant_controller and packagevariant_controller_test.

efiacor replied:
Good as is. Ignore

efiacor commented on controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go:
```diff
@@ -429,9 +487,20 @@ func (r *PackageVariantReconciler) findAndUpdateExistingRevisions(ctx context.Co
+			if err := r.Client.Update(ctx, downstream); err != nil {
+				return nil, err
+			}
 			if err := r.updatePackageResources(ctx, prr, pv); err != nil {
```
Should this be done after both the gates and condirion updates succeed?

JamesMcDermott replied:
Which is "this", the updatePackageResources or the readiness info update? (can't tell which line you're referring to, sorry... probably github). Discuss it together, maybe?

efiacor replied:
Never mind. The flow is correct. We update the gate and condition, update the pkgResources then set the condition to Ready.

efiacor commented on controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go:
```diff
@@ -694,10 +763,18 @@ func (r *PackageVariantReconciler) updateDraft(ctx context.Context,
+		return nil, err
+	}
+
+	setPrStatusCondition(draft, ConditionPipelinePVRevisionReady)
```
Not sure about this block. Do we just go ahead and set it to Ready without checking it with a get? Or is that not necessary?

JamesMcDermott replied:
Good catch... I've added a Reader.Get everywhere we set it to Ready. I *want* to say it's not necessary, but it could just be we haven't had a complicated enough package to test it with to make it run into optimistic-locking troubles.

efiacor commented on pkg/engine/engine.go:
```diff
@@ -129,63 +129,64 @@ func (cad *cadEngine) ListPackageRevisions(ctx context.Context, repositorySpec *
 }
 
-func (cad *cadEngine) CreatePackageRevision(ctx context.Context, repositoryObj *configapi.Repository, obj *api.PackageRevision, parent repository.PackageRevision) (repository.PackageRevision, error) {
+func (cad *cadEngine) CreatePackageRevision(ctx context.Context, apiRepo *configapi.Repository, newPkgRev *api.PackageRevision, parent repository.PackageRevision) (repository.PackageRevision, error) {
```
Why apiRepo?

JamesMcDermott replied:
That's to indicate that this is the Repository object from the Kubernetes API side - repositoryObj doesn't really say one way or the other and it was quite confusing given it's used later on to get a repository.Repository called "repo".
(suggestions for a better name for "repo" also welcome :))

efiacor replied:
Maybe configApiRepo? You are right that a lot of the var names are not clear to the reader.

efiacor commented on pkg/externalrepo/git/draft.go:
```diff
@@ -52,11 +52,11 @@ func (d *gitPackageRevisionDraft) Key() repository.PackageRevisionKey {
 	defer span.End()
 
-	return d.repo.UpdateDraftResources(ctx, d, new, change)
+	return d.repo.UpdateDraftResources(ctx, d, new, task)
```
changeTask?

JamesMcDermott replied:
Sounds good - done.

efiacor commented on pkg/task/kio.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022, 2024 The kpt and Nephio Authors
+// Copyright 2022, 2024-2025 The kpt and Nephio Authors
```
Necessary?

JamesMcDermott replied:
Not any more - reverted.

efiacor commented on pkg/task/taskhandler.go:
```diff
@@ -28,6 +28,10 @@ import (
 var tracer = otel.Tracer("task")
 
+var DefaultReadinessConditions = []api.Condition{
+	ConditionPipelineNotPassed,
```
Silly question. If a pkg doesn't have a pipeline, does it still get this condition added?

JamesMcDermott replied:
Short answer: yes (at least at the point of init, clone, or edit). Should it? Good question.
This condition might need renaming, and at least in function it probably overlaps the render status work - can we discuss?
Is it reasonable to suppose that a pipeline already in a Kptfile will be run by kpt in the process of rendering, and that there's no way to distinguish the pipeline's result from the rest of the render?

efiacor replied:
Is there a way to check if the KptFile does actually have a pipeline, and only then add the default condition?
Really not sure on the latter part of your comment.

efiacor commented on test/e2e/e2e_test.go:
```diff
@@ -27,9 +27,13 @@ import (
 	porchapi "github.com/nephio-project/porch/api/porch/v1alpha1"
 	configapi "github.com/nephio-project/porch/api/porchconfig/v1alpha1"
+
+	variantapi "github.com/nephio-project/porch/controllers/packagevariants/api/v1alpha1"
```
pvapi?

JamesMcDermott replied:
That works better - changed it here and in suite.go and suite_utils.go.

lapentad commented:
Merge conflicts to be fixed, but I tested locally the package resources handling and the Git conflict are better highlighted and easier to understand.

JamesMcDermott committed:
changes from review comments

JamesMcDermott committed:
regenerated mock_PackageRevision

JamesMcDermott committed:
Issue #615 - reworked in response to review comment

- removed PackagePipelinePassed from default conditions
  - left machinery in place for default conditions/readiness gates
    for future use
- now setting PackagePipelinePassed only at point of pipeline render
  - determined by:
    - presence of pipeline element in Kptfile, AND
    - presence of rendering task (type == "render" or image == "render")
      in package revision's task list

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
fix test data for CLI E2E tests

JamesMcDermott committed:
Issue #615 - continued reworking to fix issues from Nephio-level tests

- was dependent on debugging to find and fix an issue in a test artefact
  - see https://github.com/nephio-project/catalog/pull/110
- resolved performance degradation in packagevariant_controller
  - added too many calls to re-read package revisions or resources
    - should have it pared down to the minimum to work OK
- detection of pipeline now works as intended to set the "waiting
  for pipeline run" readiness condition only if the package has a
  pipeline that will do anything in the render

https://github.com/nephio-project/nephio/issues/615

JamesMcDermott committed:
missed a couple of compile failures

JamesMcDermott committed:
Merge branch 'nephio-project:main' into pv-pipeline-readiness-gates

JamesMcDermott committed:
revert Makefile change

JamesMcDermott commented:
Other work has become higher-priority - putting this issue on hold. I've pushed the additional changes that have it working (at least for the free5gc tests).

### Remaining items to do:
    - Debug/investigate:
      - in some cases when switching the package variant's readiness condition (PVOperationsComplete), porch-server doesn't give the Git patch the right custom task type/commit message
        - minor issue, but makes flow less comprehensible from downstream's Git history
        - need to debug - is there some place where a patch task/mutation is created where we don't check if it's setting the readiness gate?
          - or some issue in the Boolean logic in the UpdateOnlySetsReadinessConditions checked?
      - issue where Porch forgets package revisions
        - appears in free5gc tests, interfering with package variants creating revisions, not sure about other tests
        - seems to occur intermittently for different catalog repos on GitHub, not sure about others
        - deleting and recreating affected repo seems to bring back the revisions
          - but may result in package variants picking up erroneous changes and creating new package revisions
        - something in packagerev metadata handling? am I correct in thinking everything existential goes through packagerevs?
        - don't think it's my changes as I didn't do anything involving caches or metadata
      - issue when updating package variant in conjunction with long-running kpt pipeline
        - appears that:
          - when packagevariant_controller updates resources to patch in the new stuff
            - server renders the pipeline, hanging controller while it waits
              - renders multiple times because of the 
          - meanwhile, more reconcile attempts... begin? are scheduled? Kubernetes stuff...
            - and all pile up behind the waiting one
              - with out-of-date resources? How?!
            - and when the waiting one finishes they all go at once
              - which is fine as long as they don't have any differing data
              - and if one does, it pushes it into the package revision resources
                - resulting in more renders
                - and near-livelock?
                  - busywork for porch-server, anyway
        - Will hopefuilly be alleviated by async work - less chance of an update getting stuck
          - and we can make packagevariant_controller check the revision's readiness conditions
          - if it has a pipeline condition set to False, kick the can down the road to the next reconcile
    - Fix:
      - golangci-lint issues - should be simple relative to the others
      - failing unit tests (make test) - should be mostly re-aligning test data, but a couple of panics
      - failing E2E tests (make test-e2e) - currently 1 failure; may be just a flake
      - failing E2E CLI tests (make test-e2e-cli) - align expected test responses for propose behaviour where it lists unmet readiness conditions
      - Sonar issues, if present - currently no data as the Sonar scan relies on running the (currently-broken) unit tests
      - Note: aim for minimal additional code changes! Test changes only wherever possible!
    - Add:
      - additional E2E tests to check readiness gates on packages with and without pipelines
    - Run:
      - all automated runs on the PR
        - and repeat locally to be sure!
      - a simple package variant against a local environment
        - and confirm Git history in repo with resulting downstream package revision contains all readiness-gate/condition switches
          - PVOperationsComplete coming from packagevariant_controller (included in initial clone and switched to true after clone and patch)
          - PackagePipelinePassed before and after running pipeline (render-type task/mutation/commit)
      - test-infra free5gc and oai suites
        - one last time to be absolutely sure not broken or degraded again
    - Final review from community
      - ensure most recent comments from @efiacor (here) and @kispaljr (offline) are addressed (and responded to)

JamesMcDermott commented:
/hold

JamesMcDermott committed:
Merge branch 'nephio-project:main' into pv-pipeline-readiness-gates
